OpenRA.Mods.Common/Traits/Render/WithBuildingRepairDecoration.cs

```diff
@@ -22,24 +24,30 @@ public class WithBuildingRepairDecorationInfo : WithDecorationInfo, Requires<Rep
 
 	public class WithBuildingRepairDecoration : WithDecoration
 	{
-		readonly RepairableBuilding rb;
+		readonly RepairableBuilding[] repairableBuildings;
 		readonly WithBuildingRepairDecorationInfo info;
 		int shownPlayer = 0;
 
 		public WithBuildingRepairDecoration(Actor self, WithBuildingRepairDecorationInfo info)
 			: base(self, info)
 		{
 			this.info = info;
-			rb = self.Trait<RepairableBuilding>();
-			anim = new Animation(self.World, info.Image,
-				() => !rb.RepairActive || rb.IsTraitDisabled || !ShouldRender(self));
-			anim.PlayThen(info.Sequence, CycleRepairer);
+			repairableBuildings = self.TraitsImplementing<RepairableBuilding>().ToArray();
+
+			Func<bool> pausedFunc = () =>
+			{
+				var rb = repairableBuildings.FirstEnabledTraitOrDefault();
```

Comment:
Do we want to use `FirstEnabledTraitOrDefault` everywhere? We already know that the trait is conditional so it doesn't make sense to check if it is conditional

Reply:
What do you mean with checking if the trait is conditional? `FirstEnabledTraitOrDefault` is returning the currently active (enabled) instance of the trait. That instance can change when the conditions change, so we can't cache it easily.

Reply:
I meant this
<img width="640" alt="Screenshot 2022-10-09 at 20 31 41" src="https://user-images.githubusercontent.com/37534529/194771039-73aff2f5-83bb-4163-938f-58ca1455c0b1.png">

We know in advance that `RepairableBuilding` inherits `IDisabledTrait` so there's no need to cast. I seems this is misused more widely. I dunno if this concern is worth anything as I don't know how expensive casting is, but it's just an observation I had. (Misused everywhere where we use this on a class, for interfaces this function makes sense as they would need to cast regardless)

Reply:
Good idea. I filed #20361 as an attempt to address this. I'll update either PR depending on which gets merged first.